### ISSUE TITLE
RFC Invenio "no strings attached"

### ISSUE THREAD

**Issue by greut** [inveniosoftware/invenio#1806]
**tl;dr:** invenio should be a library (with optional and external modules)

Currently invenio is a rather complicated to install software into which an
overlay is installed. As long as the overlay is minimalist it’s just fine but
you quickly start hitting some walls (and duplications). Here is an attempt to
make the overlay the first class-citizen and let invenio be a simple library.

I'll describe how the documentation could look like in this new paradigm.
## How to install invenio 2.0

All you need is `pip install invenio`. If it's nice, it doesn’t do much as long as no applications configure or use it. It's time to create your first application using invenio (we also call them _overlay_).
### Creating an application using invenio

Jump into a new directory, ideally within a new [virtualenv](https://virtualenv.pypa.io/en/latest/) but that's up to you, really. And create a `setup.py` file which is the description file for a python package (the same you find on the _cheeseshop_ (aka [PyPI](https://pypi.python.org/pypi))).

``` python
from setuptools import setup

setup(
    name='My Overlay',
    version='0.1a0',
    url='http://www.example.org',
    author='John Doe',
    author_email='john.doe@example.org',
    description='Sample application for invenio',
    install_requires=[
        'Invenio>=2',
    ],
    entry_points={
        'invenio.config': [
            'myapp = myapp.config'
        ]
    }
)
```

Then install it:

```
$ pip install -e .
```

To use a more specific version of Invenio, it's recommended to specify it via a `requirements.txt` file this way. Here, we want to use the `pu` branch.

```
-e git://github.com/inveniosoftware/invenio@pu#egg=Invenio

-e .
```

And then install it:

```
$ pip install -r requirements.txt
```

Both are identical regarding the overlay. The first one will pick invenio from PyPI while the second one will use the `pu` branch from Gitbub. The bleeding edge as they say.
### Configuration

As you’ve seen above, we defined an entry_point for `invenio.config`. It points to a module that will contain our configuration. So create your application.

```
/myapp/
|     |- __init__.py
|     `- config.py
|
|- requirements.txt
`- setup.py
```

Put the required configuration into `config.py`.

``` python
CFG_SITE_LANGS = ["en"]

CFG_SITE_NAME = "My Overlay"
CFG_SITE_NAME_INTL = {
    "en": "My Overlay"
}

PACKAGES = [
    "myapp.base",
    "myapp.modules.*",
    "invenio.modules.*",
]
```
### Sensitive configuration

Other configuration elements like database username and password or the website url should not be put here as this file is not specific to the installation and may be put under a version control system such as Git or Subversion.

The configuration can be handled via the `inveniomanage` command line interface (or by editing the `invenio.cfg` file in the instance folder and reloading the application).

```
# boilerplate
$ inveniomanage config set create secret-key
# MySQL configuration
$ inveniomanage config set CFG_DATABASE_NAME mysql-database
$ inveniomanage config set CFG_DATABASE_USER mysql-user
$ inveniomanage config set CFG_DATABASE_PASS mysql-password
# HOST configuration (for redirects, etc.)
$ inveniomanage config set CFG_SITE_URL http://0.0.0.0:4000/
$ inveniomanage config set CFG_SITE_SECURE_URL https://0.0.0.0:4000/
```
### The base module

The `myapp.base` module is the module in which you’ll be able to override the invenio layout (templates), add views, etc. Let’s create a minimal one.

```
/myapp/base/
|     |    |- __init__.py
|     |    |- static
|     |    |- templates
|     |    `- views.py
|     |
|     |- __init__.py
|     `- config.py`
|
|- requirements.txt
`- setup.py
```

`static` and `templates` are empty directories for now. Let’s register our views with a new _Blueprint_.

``` python
from flask import Blueprint

blueprint = Blueprint('myapp', __name__, url_prefix="/",
                      template_forlder="templates", static_folder="static")
```

Et voilà!
### JavaScript and CSS

Invenio comes with HTML templates, JavaScript and CSS files that are ready to use. Some of them are using some open source libraries that aren’t yet installed. For that we recommend to use `bower` and start with the `bower.json` files provided by Invenio.

**NOTE** the `bower.json` file from invenio is the minimum set of requirements to use the default user interface of Invenio. We recommend you to built atop of it.

First thing first, bower is needed. It’s a package management tool that can be downloaded using npm another package management tool that can also be downloaded using your favourite package management tool such as `apt` or `yum`. So meta, such package management.

```
$ # Recommended way for Ubuntu LTS
$ sudo apt-get install software-properties-common
$ sudo add-apt-repository ppa:chris-lea/node.js
$ sudo apt-get update
$ sudo apt-get install nodejs
$ sudo su -c "npm install -g bower"
```

Invenio uses a little tool called `Flask-Collect` to grab all the static files used by every modules and copy them into one safe place. That is important because we want to put the downloaded assets into such a place, to make things easy.

By convention, we are using `/vendors` for all the external libraries. By default bower downloads into a folder called `bower_components`. What we want here is to download into `myapp/base/static/vendors` instead. This way when the collection is down, those file will get copied over and be accessible from the `/vendors` prefix. To do so, create a file called `.bowerrc` at the top level of the project (next to `setup.py`)

``` json
{
    "directory": "myapp/base/static/vendors"
}
```

Now, let's download all the assets from bower and collect them into the
instance directory.

``` console
$ bower install
$ inveniomanage collect
```

And magically, the `vendors` directory gets updated.

**TIP:** don't put the vendors directory into version control.

The assets will be preprocessed (for LESS or SASS), merged and even minified (clean-css or uglifyjs) by webassets which requires those external tools.

```
$ sudo su -c "npm install -g clean-css less uglifyjs requirejs"
```

All set!
### Installing the database

```
$ inveniomanage database --init --user=root --password=<ROOT> --yes-i-know
$ inveniomanage database create
```
### Go for it.

```
$ inveniomanage runserver
```
## Why is it better?

As soon as you start making something big with your overlay, you end up redoing a Gruntfile and bower.json (see 1. and 2.) or something else if it floats your boat. The CDS overlay (see 3.) is using this _alternative_ way.
1. https://github.com/inspirehep/inspire-next/tree/new-templates
2. https://github.com/zenodo/zenodo
3. https://github.com/greut/cds-demosite/tree/prototype

This way invites you to use Bower, but you can remove it totally if you want to tweak the web assets used by the bundles (currently not easy to do).

As an invenio user all that matters should be your overlay, and deploying it should be as easy as, [Deploying with Fabric](http://flask.pocoo.org/docs/patterns/fabric/):

```
$ fab pack deploy
```
### How modular can Invenio be?

Everything in `invenio.modules` is (potentially) a module. Most of the module could (and probably should) live in their own repository and be part of the dependencies (`install_requires`) of the overlay.
- **+** the upside is that testing can be confined to a module avoiding Travis timeout (#1786) and potential tests interferences.
- **-** The downside is that the documentation will be fragmented.
## What should be changed?
- Grunt is no longer required delegating all the heavy work to webassets.
  See: https://forum.invenio-software.org/t/webassets-vs-grunt/45
- A default bower file should be generated just like the Apache conf is.
- Bundles must be defined in the modules themselves and be easily modifiable
  from the overlay.
  See: #1767
